src/metabase/db/liquibase.clj

```diff
@@ -228,14 +232,52 @@
   [liquibase]
   (.hasChangeLogLock (lock-service liquibase)))
 
+(defn- wait-until [done? ^long sleep-ms timeout-ms]
+  (let [deadline (+ (System/nanoTime) (* 1e6 timeout-ms))]
+    (loop []
+      (if (done?)
+        :done
+        (if (>= (System/nanoTime) deadline)
+          :timed-out
+          (do (Thread/sleep sleep-ms)
+              (recur)))))))
+
+(defn- locked-instances []
+  (filter holding-lock? @potentially-locked-instances))
+
+(defn wait-for-all-locks
+  "Wait up to a maximum of `timeout-seconds` for the given Liquibase instance to release the migration lock."
+  [sleep-ms timeout-ms]
+  (log/warn (locked-instances))
+  (let [done? #(empty? (locked-instances))]
+    (if (done?)
+      :none
+      (do (log/info "Waiting for migration lock(s) to be released")
+          (wait-until done? sleep-ms timeout-ms)))))
+
+(defn- liquibase->url [^Liquibase liquibase]
+  ;; Need to this cast to get access to the metadata. We currently only use JDBC app databases.
+  (let [conn ^JdbcConnection (.getConnection (-> liquibase (.getDatabase)))]
+    (-> conn .getMetaData .getURL)))
+
+(defn release-concurrent-locks!
+  "Release any locks held by this process corresponding to the same database."
+  [conn-or-data-source]
+  (when-let [instances (not-empty (locked-instances))]
+    (with-liquibase [liquibase conn-or-data-source]
+      (let [url (liquibase->url liquibase)]
+        (doseq [instance instances]
+          (when (= url (liquibase->url instance))
```

Comment:
There's a bit of an unfortunate story here.

The liquibase namespace is agnostic as to what databases it is run against, so it's technically possible its running migrations against multiple databases - it doesn't know about our single global app db architecture, and I don't think it should - needless coupling.

In order to release the lock, we need a separate connection to the one that's tied up performing the migrations, and there's no way to "clone" the existing connection as we're unable to recover the password. For this reason we need to pass the db connection details in from our db.setup namespace.

Since we're now passing in the details for only a specific DB, I feel that we should ignore any of the (theoretical) other databases under migration. If it's only other dbs being modified there's nothing we can do about them as we don't have their connection details.

Reply:
We chatted about this offline, and decided it's worth replicating the "connection deadlock" issue this is hacking around, ideally captured as a test.

Reply:
So I didn't find a good way to replicate this in a test, without blowing up the test database with a long running query. I did manage to reproduce the blocking behavior by running a dev server from the CLI and sending that an actual SIGTERM though. With this "fresh" instance trick the blocking is avoided.